metals/src/main/scala/scala/meta/internal/metals/debug/SocketEndpoint.scala

```diff
@@ -3,22 +3,34 @@ package scala.meta.internal.metals.debug
 import java.io.BufferedInputStream
 import java.io.BufferedOutputStream
 import java.net.Socket
-import java.util.Collections
 
-import scala.meta.internal.metals.debug.SocketEndpoint._
+import scala.collection.concurrent.TrieMap
 
+import org.eclipse.lsp4j.debug.services.IDebugProtocolServer
 import org.eclipse.lsp4j.jsonrpc.MessageConsumer
 import org.eclipse.lsp4j.jsonrpc.debug.json.DebugMessageJsonHandler
+import org.eclipse.lsp4j.jsonrpc.json.MethodProvider
 import org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer
 import org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer
 import org.eclipse.lsp4j.jsonrpc.messages.Message
+import org.eclipse.lsp4j.jsonrpc.messages.RequestMessage
+import org.eclipse.lsp4j.jsonrpc.services.ServiceEndpoints
 
 private[debug] final class SocketEndpoint(socket: Socket)
     extends RemoteEndpoint {
+  private val methods =
+    ServiceEndpoints.getSupportedMethods(classOf[IDebugProtocolServer])
+  private val handler = new DebugMessageJsonHandler(methods)
   private val source = messageSource(socket)
   private val target = messageTarget(socket)
+  private val ongoingRequests = TrieMap.empty[String, String]
 
   override def consume(message: Message): Unit = {
+    message match {
+      case msg: RequestMessage =>
+        ongoingRequests.put(msg.getId(), msg.getMethod())
```

Comment:
We also need to pair a request-id with a request to know how to deserialize each body.